TOSA legalization updates for spec v0.22, Part 1

sjarus opened the pull request:
- Updated gather, gather_nd and resize.
- Fix precision of TFL avgpool2d, quantize. Add squareddifference
- Add left/right shift, leaky_relu, one_hot
- Update relu6/relu_n1_to1
- Numerical precision in tf.fakequant legalization

Signed-off-by: Suraj Sudhir <suraj.sudhir@arm.com>
Change-Id: Idf23c3b24342f75ee7d1eb22dab6ffe27e8710b1

sjarus committed:
Collection of TOSA legalization updates
- Updated gather, gather_nd and resize.
- Fix precision of TFL avgpool2d, quantize. Add squareddifference
- Add left/right shift, leaky_relu, one_hot
- Update relu6/relu_n1_to1
- Numerical precision in tf.fakequant legalization

Signed-off-by: Suraj Sudhir <suraj.sudhir@arm.com>
Change-Id: Idf23c3b24342f75ee7d1eb22dab6ffe27e8710b1

sjarus commented:
@stellaraccident and @rsuderman , here are the part 1 set of updates to legalizations from TF/TFLite to TOSA aligned to the recent LLVM side update. That change was picked up by TensorFlow yesterday so we're pushing this out.

sjarus commented on tensorflow/compiler/mlir/tosa/BUILD:
```diff
@@ -79,6 +79,7 @@ cc_library(
         "//tensorflow/compiler/mlir/lite:tensorflow_lite",
         "//tensorflow/core:framework",
         "//tensorflow/core/kernels:conv_grad_shape_utils",
+        "//tensorflow/core/kernels:fake_quant_ops",
```
It implements tensorflow::Nudge that is invoked from convertFakeQuantOp() in legalize_common.cc

stellaraccident replied:
Thanks - i will try to review in detail tonight but have a lot in my queue and it may slip to tomorrow.

stellaraccident replied:
This will add a cross dependency on a large body of TensorFlow to the TFLite path (est. 1000 additional source files?). I would rather just copy the Nudge function locally. (this has been done a couple of times for dependency reasons already)

armkevincheng replied:
sure we can do that and remove the call to Nudge.

armkevincheng replied:
Thanks for suggestion. Will try to replace it with a local implementation in next round.

stellaraccident commented on tensorflow/compiler/mlir/tosa/transforms/legalize_tfl.cc:
```diff
@@ -1035,8 +1044,30 @@ LogicalResult ConvertTFLAveragePool2DOp::matchAndRewrite(
+  if (input_type.getElementType().isa<UniformQuantizedType>()) {
+    // Search for attribute annotated by --tosa-convert-tfl-uint8 pass.
+    // This is needed in average pool2d to match TFLite rounding behavior.
+    IntegerAttr override_zeropoint_attr =
```
These kind of cross-pass action at a distance attribute dependencies need to be avoided. Why can't you have a helper which performs the check in the pattern which sets this attribute and just call it here (i.e. a static `calculateZeropointOverride(TFL::AveragePool2DOp op)`).

Even better you could define the above to return an `IntegerAttr` conditionally if the input_type is quantized (or nullptr otherwise), then collapse all of the branching in this transform.

armkevincheng replied:
I agree this kind of cross-pass design is bad, but that was the only way I thought I can achieve what I want.

Briefly explain what I'm doing here:
We're trying to match the rounding behavior to TFL::AveragePool2DOp, which has different rounding behavior between QI8 and QU8 (since it's rounding on its storage type/range). But in the pass pipeline, the convert_tfl_uint8 pass converts all the QU8 into QI8, so the pass after it, e.g. legalize_tf and legalize_tfl, can't distinguish between those two cases. And that's why you see the override_zeropoint attribute is set in convert_tfl_uint8.

The trick we played here (which is bad I agree) is to annotate the zeropoint attribute based on if it's QU8 or QI8 before QU8 is converted. And in legalize_tfl here, it checks if such attribute exists. If it does, then we override that zeropoint with what's stored in the attribute, and we build TOSA::AveragePool2dOp with it.

Is the helper function you mentioned something that could solve the problem without using the cross-pass design?

armkevincheng replied:
Explained in the comment on top. QU8 and QI8 is not distinguishable after convert_tfl_uint8 pass.

sjarus replied:
This piece appears to be the largest blocker for this update. We're aware - as Kevin mentioned - that the approach isn't ideal. Is it ok if we get all other changes done, then do a separate update with a cleaner implementation of this cross pass issue ?

stellaraccident commented on tensorflow/compiler/mlir/tosa/transforms/legalize_tfl.cc:
```diff
@@ -2267,6 +2298,52 @@ LogicalResult ConvertTFLPadOp::matchAndRewrite(
+  // Not a ranked tensor output
+  if (!output_type) return failure();
+
+  bool align_corners = tfl_resize_op.align_cornersAttr().getValue();
```
Nit: I don't think these separate bools are carrying their weight. I would just inline them into the call to `convertResizeOp`.

armkevincheng replied:
Thanks will do. I'd assume same comment for "half_pixel_centers" 1 line below?

stellaraccident commented on tensorflow/compiler/mlir/tosa/transforms/legalize_tfl.cc:
```diff
@@ -2267,6 +2298,52 @@ LogicalResult ConvertTFLPadOp::matchAndRewrite(
+  // Not a ranked tensor output
+  if (!output_type) return failure();
+
+  bool align_corners = tfl_resize_op.align_cornersAttr().getValue();
```
Nit: I don't think these separate bools are carrying their weight. I would just inline them into the call to `convertResizeOp`.

armkevincheng replied:
Thanks will do. I'd assume same comment for "half_pixel_centers" 1 line below?

stellaraccident commented on tensorflow/compiler/mlir/tosa/transforms/legalize_tfl.cc:
```diff
@@ -2671,13 +2748,116 @@ LogicalResult ConvertTFLPReluOp::matchAndRewrite(
+                              rewriter.getIntegerType(1)),
+        tfl_leakyrelu_op.input(), const_zero);
+
+    auto op3_select = rewriter.create<tosa::SelectOp>(
```
Nit: Prefer that each branch of the if just end with a `rewriter.replaceOpWithNewOp<...>(...)` and a `return success();` (versus ending with a shared `replaceOp` and return after the condition).

armkevincheng replied:
Thanks. Will fix in next round.

stellaraccident commented on tensorflow/compiler/mlir/tosa/transforms/legalize_tfl.cc:
```diff
@@ -2752,21 +2932,25 @@ LogicalResult ConvertTFLQuantizeOp::matchAndRewrite(
+      qtype.getElementType().dyn_cast<UniformQuantizedType>();
   if (!element_type) return failure();
 
+  // In some cases output_type is dynamic shape as tensor<*xelement_type>
```
s/dynamic shape/unranked tensor/

stellaraccident replied:
This comment does not match what you are doing: If it is an `UnrankedTensor`, then the `input_type.getShape()` will assert. Why not just `auto output_type = input_type;` if you can afford to do that? If it needs to be ranked, then you will need to fail the pattern if it is not.

armkevincheng replied:
I remember in one of the failing full networks run, I've seen something like:
%1 = tfl.quantize(%0: tensor<4x4xf32>) {...} -> tensor<*x!uniform.quant<...>>

In which case, input shape is determined but output shape is not.

But I can't remember which test trigger this. I'll need to dig that out, and check if the shape-inference pass is called as expected.

rsuderman replied:
This kinda of case usually occurs when tensorflow's shape propagation is not working correctly. Passes should assume valid shape propagation as it is an assumed-correct upstream behavior.

stellaraccident commented on tensorflow/compiler/mlir/tosa/transforms/legalize_tf.cc:
```diff
@@ -1653,6 +1661,52 @@ LogicalResult ConvertTFPadOp::matchAndRewrite(Operation* op,
+  // Not a ranked tensor output
+  if (!output_type) return failure();
+
+  bool align_corners = tf_resize_op.align_cornersAttr().getValue();
```
nit: inline into call site?

armkevincheng replied:
Thanks will do. I'd assume same comment for "half_pixel_centers" 1 line below?

stellaraccident commented on tensorflow/compiler/mlir/tosa/transforms/legalize_tf.cc:
```diff
@@ -1653,6 +1661,52 @@ LogicalResult ConvertTFPadOp::matchAndRewrite(Operation* op,
+  // Not a ranked tensor output
+  if (!output_type) return failure();
+
+  bool align_corners = tf_resize_op.align_cornersAttr().getValue();
```
nit: inline into call site?

armkevincheng replied:
Thanks will do. I'd assume same comment for "half_pixel_centers" 1 line below?

stellaraccident commented on tensorflow/compiler/mlir/tosa/transforms/legalize_tf.cc:
```diff
@@ -1815,9 +1927,50 @@ LogicalResult ConvertTFLeakyReluOp::matchAndRewrite(
+
+  auto a2_ge = rewriter.create<tosa::GreaterEqualOp>(
+      op->getLoc(),
+      RankedTensorType::get(output_type.getShape(), rewriter.getIntegerType(1)),
```
nit: `rewriter.getI1Type()`

armkevincheng replied:
Thanks. Will fix in next round.

stellaraccident commented on tensorflow/compiler/mlir/tosa/transforms/legalize_tf.cc:
```diff
@@ -1815,9 +1927,50 @@ LogicalResult ConvertTFLeakyReluOp::matchAndRewrite(
+    alpha = tmpAttr.getValueAsDouble();
+  }
+
+  Value const_zero = getTosaConstTensorSingleF32(rewriter, op, 0.0);
```
I'm a bit concerned about the implicit assumption of f32 here (which is not an invariant of the tf op). What is the story for other data types (I believe if it happened, it would fail tosa verification, which is better than a crash, but still not great to be applying a pattern that produces a broken op).

armkevincheng replied:
Maybe I should check if data type is f32 and return failure() is it's not now. Does that sound okay for now?

rsuderman replied:
Sounds fine.

stellaraccident commented on tensorflow/compiler/mlir/tosa/transforms/legalize_common.h:
```diff
@@ -30,6 +30,7 @@ limitations under the License.
 
 namespace mlir {
 namespace tosa {
+constexpr char kOverrideZeropointAttrName[] = "override_zeropoint";
```
I think you need to remove this attribute, but if not and you kept it, you should be using a namespaced attribute (i.e. `tosa._override_zeropoint`). But please, per other comments, I think we should eliminate the use entirely.

armkevincheng replied:
Agree that I should keep this under tosa namespace. And let's discuss main concern in the top thread.

stellaraccident commented on tensorflow/compiler/mlir/tosa/transforms/legalize_common.cc:
```diff
@@ -2212,61 +2213,53 @@ llvm::Optional<Value> convertFusedActivation(PatternRewriter& rewriter,
+          op->getLoc(), input_type, input_value,
+          rewriter.getI64IntegerAttr(quantized_0),
+          rewriter.getI64IntegerAttr(quantized_max),
+          rewriter.getF32FloatAttr(0), rewriter.getF32FloatAttr(0));
```
per some other comments: what happens if `input_type` is not f32? (here and elsewhere). I feel you may be missing a helper that gets a type appropriate float attr given an input type.

armkevincheng replied:
I agree we should do more check where we implicitly assume types are f32.

But isn't this function already did the check you're asking?
if (input_is_qtype) {
  ... // legalize for quantized type
} else {
   ... // legalize for f32
   // maybe I should check and return failure() if not input_type is f32 here
}

stellaraccident commented on tensorflow/compiler/mlir/tosa/transforms/legalize_common.cc:
```diff
@@ -2577,6 +2566,197 @@ llvm::Optional<Value> convertReduceMeanOp(
+
+  auto input_shape = input_type.getShape();
+  auto output_shape = output_type.getShape();
+
```
Please verify that the input and output shape are rank 3. A comment as to what the dims are (symbolically) wouldn't hurt. I assume [batch, height, width]?

armkevincheng replied:
Yes I can add the check in next round.

Isn't NHWC, which is rank 4, being the default format for resize's input/output?

stellaraccident commented on tensorflow/compiler/mlir/tosa/transforms/legalize_common.cc:
```diff
@@ -2577,6 +2566,197 @@ llvm::Optional<Value> convertReduceMeanOp(
+  auto input_shape = input_type.getShape();
+  auto output_shape = output_type.getShape();
+
+  size_t input_height = input_shape[1];
```
I recommend making these typed `int64_t`, which is the actual range of the shape components (size_t is not 64 bits on 32 bit platforms). Recommend using a signed type because you are just feeding into signed arithmetic below.

armkevincheng replied:
Thanks. Will update in next round.

stellaraccident commented on tensorflow/compiler/mlir/tosa/transforms/legalize_common.cc:
```diff
@@ -2577,6 +2566,197 @@ llvm::Optional<Value> convertReduceMeanOp(
+  size_t output_height = output_shape[1];
+  size_t output_width = output_shape[2];
+
+  bool input_is_qtype =
```
nit: move these checks to the top, keeping the height/width extraction with the calculations that use them.

armkevincheng replied:
Thanks. Will update in next round.

stellaraccident commented on tensorflow/compiler/mlir/tosa/transforms/legalize_common.cc:
```diff
@@ -2577,6 +2566,197 @@ llvm::Optional<Value> convertReduceMeanOp(
+        rewriter.getF32ArrayAttr({fp_offset_y, fp_offset_x}), resize_mode);
+    val = resize_op.getResult();
+  }
+  return val;
```
nit: prefer to return in each branch of the condition above vs joining and then returning an intermediate.

armkevincheng replied:
Thanks. Will update in next round.

stellaraccident commented on tensorflow/compiler/mlir/tosa/transforms/legalize_common.cc:
```diff
@@ -2797,5 +2980,563 @@ llvm::Optional<Value> convertTFConv2DCommon(
+                                      Value indices_value, int32_t batch_dims,
+                                      int32_t axis)
+
+{
```
nit: remove newline and put opening brace inline with function defn.

armkevincheng replied:
Thanks. Will update in next round.

stellaraccident commented on tensorflow/compiler/mlir/tosa/transforms/legalize_common.cc:
```diff
@@ -2797,5 +2980,563 @@ llvm::Optional<Value> convertTFConv2DCommon(
+  auto params_type = params_value.getType().dyn_cast<RankedTensorType>();
+  auto indices_type = indices_value.getType().dyn_cast<RankedTensorType>();
+
+  if (!result_type) return llvm::None;
```
nit: one if statement with multiple || clauses for compactness.

armkevincheng replied:
Thanks. Will update in next round.

stellaraccident commented on tensorflow/compiler/mlir/tosa/transforms/legalize_common.cc:
```diff
@@ -2797,5 +2980,563 @@ llvm::Optional<Value> convertTFConv2DCommon(
+
+  int32_t N = 1, W = 1, K = 1, C = 1;
+
+  size_t params_rank = params_type.getShape().size();
```
`params_type.getRank()`

armkevincheng replied:
Thanks. Will update in next round.

stellaraccident commented on tensorflow/compiler/mlir/tosa/transforms/legalize_common.cc:
```diff
@@ -2797,5 +2980,563 @@ llvm::Optional<Value> convertTFConv2DCommon(
+
+  int32_t N = 1, W = 1, K = 1, C = 1;
+
+  size_t params_rank = params_type.getShape().size();
```
please use `int` for these ranks. (it is range appropriate and better to not be mixing signed and unsigned arithmetic below)

armkevincheng replied:
Thanks. Will update in next round.

stellaraccident commented on tensorflow/compiler/mlir/tosa/transforms/legalize_common.cc:
```diff
@@ -2797,5 +2980,563 @@ llvm::Optional<Value> convertTFConv2DCommon(
+
+  int32_t N = 1, W = 1, K = 1, C = 1, ND = 1;
+
+  size_t params_rank = params_type.getShape().size();
```
use `int` and `params_type.getRank()`

armkevincheng replied:
Thanks. Will update in next round.

stellaraccident commented on tensorflow/compiler/mlir/tosa/transforms/legalize_common.cc:
```diff
@@ -2797,5 +2980,563 @@ llvm::Optional<Value> convertTFConv2DCommon(
+  //
+  // Where, Indices is indices.shape[0:ND-1]
+
+  int32_t N = 1, W = 1, K = 1, C = 1, ND = 1;
```
Just use `int`, not `int32_t`

armkevincheng replied:
Thanks. Will update in next round.

stellaraccident commented on tensorflow/compiler/mlir/tosa/transforms/legalize_common.cc:
```diff
@@ -2797,5 +2980,563 @@ llvm::Optional<Value> convertTFConv2DCommon(
+  // within range of [0, depth - 1] and the [..., k, ...] = on_value (if k =
+  // index), or [..., k, ...] = off_value (if k != index)
+  //
+  // The lowering below assumes depth is always known at compile time.
```
Where are you checking this condition and early exiting? (forgive me if I missed it)

armkevincheng replied:
No I didn't. I will add the check in next round.

armkevincheng replied:
Oh actually we did. In legalize_tf.cc and legalize_tfl.cc, we do return failure() if matchPattern() fails.

stellaraccident commented on tensorflow/compiler/mlir/tosa/transforms/legalize_common.cc:
```diff
@@ -2797,5 +2980,563 @@ llvm::Optional<Value> convertTFConv2DCommon(
+  for (int32_t i = 0; i < indices_type.getRank(); i++) {
+    int32_t dim = indices_type.getShape()[i];
+    N *= dim;
+    if (i >= axis)
```
Unlike LLVM style, google style wants braces around these.

armkevincheng replied:
Will add braces in next round.

sjarus replied:
Done

stellaraccident commented on tensorflow/compiler/mlir/tosa/transforms/legalize_common.cc:
```diff
@@ -2797,5 +2980,563 @@ llvm::Optional<Value> convertTFConv2DCommon(
+    axis = indices_type.getRank();
+  }
+
+  int32_t N = 1, W = 1, C = 1;
```
Prefer just `int` instead of `int32_t` unless if there is a strong reason (here and below).

armkevincheng replied:
Thanks will update in next round.

stellaraccident commented on tensorflow/compiler/mlir/tosa/tests/tf-to-tosa-pipeline.mlir:
```diff
@@ -764,22 +771,64 @@ func @test_depth_to_space(%arg0: tensor<1x32x32x8xf32>) -> tensor<1x64x64x2xf32>
+// CHECK: tosa.reshape
+// CHECK: tosa.transpose
+// CHECK: tosa.reshape
+func @test_one_hot(%arg0: tensor<4x4xi32>, %arg1: tensor<f32>, %arg2: tensor<f32>) -> tensor<4x4x2xf32> {
```
General note on tests (and specific note on this): These are non trivial lowerings and we should have more explicit tests. The tests do not need to verify every detail of the patterns but should minimally capture SSA values and constants. Getting the testing granularity right is a bit of an art, so if you want help on this, we can bust some examples out to a doc or gist or something and work on it together if that is helpful (been meaning to ask for test upgrades for some time and realize there may be some education here).

armkevincheng replied:
Ah so we can check the SSA values as well. Didn't realize we can do that.
I'd assume all we need to do is append that to "// CHECK: tosa.XXX", for example like "// CHECK: tosa.XXX(%0)"?
or it's not that simple as I expect?

rsuderman replied:
You can see similar examples in the tensorflow to hlo lowering tests:

https://github.com/tensorflow/tensorflow/blob/master/tensorflow/compiler/mlir/xla/tests/legalize-tf.mlir#L64

The problem with not validating the large sequences is incorrect constants or changing which values are passed to which ops.

rsuderman replied:
Nit: You can drop all the shape information in the line to make things more succinct.

I.e. The following line:

// CHECK: %[[VAR0:.*]] = "tosa.const"() {value = dense<[0, 2, 1]> : tensor<3xi32>} : () -> tensor<3xi32>

Can be simplified to

// CHECK: %[[VAR0:.*]] = "tosa.const"() {value = dense<[0, 2, 1]> : tensor<3xi32>}

Generally we want to make sure check tests are mostly a sequence of correct values.

stellaraccident commented on tensorflow/compiler/mlir/tosa/tests/tf-to-tosa-pipeline.mlir:
```diff
@@ -105,9 +105,16 @@ func @test_relu6(%arg0: tensor<13x21x3xf32>) -> tensor<13x21x3xf32> {
 // -----
 
```
I didn't double check, but I think that some of the op conversions added are missing tests (gather?). Please audit test coverage for both ops and critical branches in the patterns.

armkevincheng replied:
Thanks. Will fix in next round.

rsuderman replied:
Still missing tests for gather support. Because we are adding functionality for a gather lowering we will need a test to validate the lowering is working correctly.

stellaraccident reviewed (commented):
Thank you for the contributions. There are a number of comments ranging from local/style, to design to testing. I think the main design point has to do with the `override_zero_point` attribute and doing that in a way that does not introduce a cross pass dependency. The other big request is upgraded test coverage. I'm happy to jump on some kind of shared space and help educate on some practices here. In addition, I'd suggest looking at the way that some of the `legalize_tf` tests are done in the tensorflow/xla side for an idea of more of what we are looking for.

armkevincheng commented on tensorflow/compiler/mlir/tosa/transforms/convert_tfl_uint8.cc:
```diff
@@ -128,6 +128,48 @@ struct ConvertUint8QConstOp : public RewritePattern {
+    auto tfl_avgpool2d_op = cast<TFL::AveragePool2DOp>(op);
+
+    auto input_type =
+        tfl_avgpool2d_op.input().getType().dyn_cast<RankedTensorType>();
```
my bad. will fix in next round.

armkevincheng replied:
My bad. Will fix in next round.

armkevincheng commented on tensorflow/compiler/mlir/tosa/transforms/convert_tfl_uint8.cc:
```diff
@@ -128,6 +128,48 @@ struct ConvertUint8QConstOp : public RewritePattern {
+
+    // Annotate attribute to match TFLite average pool2d rounding behavior.
+    auto key =
+        mlir::Identifier::get(kOverrideZeropointAttrName, builder.getContext());
```
Thanks for suggestion. Will update in next round.

armkevincheng replied:
Thanks. Will update in next round.

armkevincheng commented on tensorflow/compiler/mlir/tosa/transforms/convert_tfl_uint8.cc:
```diff
@@ -128,6 +128,48 @@ struct ConvertUint8QConstOp : public RewritePattern {
   }
 };
 
+struct ConvertAveragePool2DOp : public RewritePattern {
```
Do you have example for this that we can dumbly follow?

armkevincheng commented on tensorflow/compiler/mlir/tosa/transforms/convert_tfl_uint8.cc:
```diff
@@ -128,6 +128,48 @@ struct ConvertUint8QConstOp : public RewritePattern {
+    auto value = builder.getI32IntegerAttr(override_zeropoint);
+    op->setAttr(key, value);
+
+    return success();
```
Please see my comment on top.

armkevincheng commented:
I agree the cross-pass design is bad, but that was the only way I thought I can achieve what I want.

Briefly explain what I'm doing here:
We're trying to match the rounding behavior to TFL::AveragePool2DOp, which has different rounding behavior between QI8 and QU8 (since it's rounding on its storage type/range).

In the TFL to TOSA pass pipeline, the convert_tfl_uint8 pass is called first, and converts all the QU8 into QI8, so the pass after, e.g. legalize_tfl, only need to deal with QI8, but at the same time can't distinguish between those two cases.

The trick we played here (which is bad I agree) is to annotate the override_zeropoint attribute based on if it's QU8 or QI8 before QU8 is converted. When pass reaches legalize_tfl, it checks if such attribute exists. If it does, then we override that zeropoint with what's stored in the attribute, and we build TOSA::AveragePool2dOp with it.

Could something you mentioned above solve the problem without using the cross-pass design?

armkevincheng commented:
Again thank for all the feedbacks. They're all pretty helpful and we'll prepare next round of review as soon as possible.

sjarus committed:
Updated with changes to PR https://github.com/tensorflow/tensorflow/pull/48193

- Use tensorflow::Nudge implementation locally to avoid build dependency
- Cleanup in coding artifacts
- Disable cross-pass AvgPool2d int8/uint8 handling until better solution

Change-Id: I1e0421ee336e4e137aebfa2856870fa3838169e6
Signed-off-by: Suraj Sudhir <suraj.sudhir@arm.com>

sjarus commented:
We've removed avgpool until the cross pass dependency is resolved, but have updated other things. We hope we can upstream this, and follow up the remaining pieces in a separate PR related to the remaining TOSA v0.22 changes still to follow.

sjarus committed:
Added tests for legalizations in https://github.com/tensorflow/tensorflow/pull/48193
Fixed tfl.quantize legalization output_type setup.

Change-Id: I4a4826f653941299916229d8e2d0e094342f8ef4
Signed-off-by: Suraj Sudhir <suraj.sudhir@arm.com>

sjarus commented:
We're updating test_one_hot to follow the suggested approach. If that looks good, we'll follow up with a clean up of the entire set of tests in a follow up PR since it is an independent task. However we intend to add the gather tests to this PR, hopefully enabling us to close this one out.

rsuderman commented on tensorflow/compiler/mlir/tosa/tests/tf-to-tosa-pipeline.mlir:
```diff
@@ -764,23 +771,103 @@ func @test_depth_to_space(%arg0: tensor<1x32x32x8xf32>) -> tensor<1x64x64x2xf32>
+// CHECK: tosa.mul
+// CHECK: tosa.reshape
+// CHECK: tosa.add
 func @test_fakequant_with_min_max_args(%arg0: tensor<13x21x3xf32>) -> tensor<13x21x3xf32> {
```
Perform the same value validation on the rest of the added tests. If they are short (e.g. less than 3-4 ops) its fine without them but most of these tests are pretty complex and we want to avoid future errors.

sjarus replied:
The plan is to clean up all the tests in a separate PR. Is that a workable option ? We tried the test_one_hot to get feedback on the right way to go about it, rather than change everything, then potentially have to do a second pass at everything.

rsuderman reviewed (changes requested):
Last two comments related to tests. Once these wrap we should be good to land.

sjarus commented:
Please let us know if we can update all the TF + TFL tests in a separate PR, using the suggested template of test_one_hot . Since this is largely independent of this PR and the ones to follow it, we can parallelize the test update work.

rsuderman reviewed (approved):
We can land under the guarantee that the followup appears reasonably soon. We tend to avoid landing tests with brittle checks in case TF canonicalizations are updated. This can cause some unexpected failures.

sjarus commented:
> We can land under the guarantee that the followup appears reasonably soon. We tend to avoid landing tests with brittle checks in case TF canonicalizations are updated. This can cause some unexpected failures.

Current estimate is for the updated tests for both TF and TFL legalization to constitute a new PR within the next week. Is that reasonable ?

rsuderman commented on tensorflow/compiler/mlir/tosa/transforms/legalize_tfl.cc:
```diff
@@ -2671,13 +2723,108 @@ LogicalResult ConvertTFLPReluOp::matchAndRewrite(
+  if (!input_type || !output_type) return failure();
 
-  return failure();
+  bool input_is_qtype =
```
Value is unused.

sjarus replied:
done

rsuderman commented on tensorflow/compiler/mlir/tosa/transforms/legalize_common.cc:
```diff
@@ -2577,6 +2599,200 @@ llvm::Optional<Value> convertReduceMeanOp(
+  // oh * fp_stride_y + fp_offset_y = ix
+
+  ArrayAttr output_size =
+      rewriter.getI64ArrayAttr({output_height, output_width});
```
output_height and output_width require explicit static_casts for the initializer list.

rsuderman commented on tensorflow/compiler/mlir/tosa/transforms/legalize_common.cc:
```diff
@@ -2577,6 +2599,200 @@ llvm::Optional<Value> convertReduceMeanOp(
+      RankedTensorType output_acc_type;
+      auto input_element_qtype =
+          input_type.getElementType().cast<mlir::quant::UniformQuantizedType>();
+      auto output_element_qtype =
```
Unused

sjarus replied:
done

rsuderman commented on tensorflow/compiler/mlir/tosa/transforms/legalize_common.cc:
```diff
@@ -2577,6 +2599,200 @@ llvm::Optional<Value> convertReduceMeanOp(
+        op->getLoc(), output_type, input_value, output_size,
+        rewriter.getI64ArrayAttr({0, 0}), rewriter.getI64ArrayAttr({0, 0}),
+        rewriter.getI32IntegerAttr(0),
+        rewriter.getF32ArrayAttr({fp_stride_y, fp_stride_x}),
```
Explicit cast to float required

sjarus replied:
done

rsuderman commented on tensorflow/compiler/mlir/tosa/transforms/legalize_common.cc:
```diff
@@ -2577,6 +2599,200 @@ llvm::Optional<Value> convertReduceMeanOp(
+        rewriter.getI64ArrayAttr({0, 0}), rewriter.getI64ArrayAttr({0, 0}),
+        rewriter.getI32IntegerAttr(0),
+        rewriter.getF32ArrayAttr({fp_stride_y, fp_stride_x}),
+        rewriter.getF32ArrayAttr({fp_offset_y, fp_offset_x}), resize_mode);
```
Same casts as above.

sjarus replied:
done

rsuderman commented on tensorflow/compiler/mlir/tosa/transforms/legalize_common.cc:
```diff
@@ -2577,6 +2599,200 @@ llvm::Optional<Value> convertReduceMeanOp(
+  double in_center_h = static_cast<double>(input_height - 1) / 2.0;
+  double in_center_w = static_cast<double>(input_width - 1) / 2.0;
+  double out_center_h = static_cast<double>(output_height - 1) / 2.0;
+  double out_center_w = static_cast<double>(output_width - 1) / 2.0;
```
All of these doubles are unused

sjarus replied:
done

rsuderman commented on tensorflow/compiler/mlir/tosa/transforms/legalize_common.cc:
```diff
@@ -2638,26 +2848,18 @@ llvm::Optional<Value> convertDequantizeOp(PatternRewriter& rewriter,
 
   auto output_shape = output_type.getShape();
```
Unused

rsuderman commented on tensorflow/compiler/mlir/tosa/transforms/legalize_common.cc:
```diff
@@ -2686,30 +2888,47 @@ llvm::Optional<Value> convertFakeQuantOp(PatternRewriter& rewriter,
 
   auto output_shape = output_type.getShape();
```
Unused.

rsuderman commented:
Found a set of changes required to fix some internal failures, mostly unused variable issues.

rsuderman commented:
Working on landing this internally patching the changes in myself.

sjarus commented:
> Working on landing this internally patching the changes in myself.

Oops didn't see this. Thanks for handling this!
